From 4abca9ce1f12173ae1db87213cccb00d912e8670 Mon Sep 17 00:00:00 2001
From: Shunqing Chen <csq@rock-chips.com>
Date: Thu, 23 Sep 2021 15:36:41 +0800
Subject: [PATCH 104/232] misc: rk628-for-all: config 4K yuv420 in dts

If src mode is 4K yuv420, add "src-mode-4k-yuv420"in dts,
otherwise hdmirx does not know the input color format.

Signed-off-by: Shunqing Chen <csq@rock-chips.com>
Change-Id: I8108fff5ad3e66751c2404ea4b7a8ab5fed3f1d1
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/misc/rk628/rk628_hdmirx.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/rk628/rk628_hdmirx.c b/drivers/misc/rk628/rk628_hdmirx.c
index 5719970b55771e..6a6de10e2ea6f0 100644
--- a/drivers/misc/rk628/rk628_hdmirx.c
+++ b/drivers/misc/rk628/rk628_hdmirx.c
@@ -75,6 +75,7 @@ struct rk628_hdmirx {
 	u32 fs_audio;
 	bool audio_present;
 	bool hpd_output_inverted;
+	bool src_mode_4K_yuv420;
 };
 
 const static struct rk628_hdmi_mode supported_modes[] = {
@@ -414,7 +415,7 @@ static int rk628_hdmirx_phy_setup(struct rk628 *rk628)
 	 * force 594m mode to yuv420 format.
 	 * bit30 is used to indicate whether it is yuv420 format.
 	 */
-	if (dst_mode->clock == 594000)
+	if (hdmirx->src_mode_4K_yuv420 && dst_mode->clock == 594000)
 		f |= BIT(30);
 
 	for ( i = 0; i < RXPHY_CFG_MAX_TIMES; i++) {
@@ -442,7 +443,7 @@ static int rk628_hdmirx_phy_setup(struct rk628 *rk628)
 			       width, height, frame_width,
 			       frame_height, status, cnt);
 
-			if (dst_mode->clock == 594000)
+			if (hdmirx->src_mode_4K_yuv420 && dst_mode->clock == 594000)
 				width *= 2;
 
 			if (cnt >= 15)
@@ -470,6 +471,8 @@ static int rk628_hdmirx_phy_setup(struct rk628 *rk628)
 			src_mode->vsync_end = hdmirx->mode.vend;
 			src_mode->vtotal = hdmirx->mode.vtotal;
 			src_mode->flags = hdmirx->mode.flags;
+			if (hdmirx->src_mode_4K_yuv420 && dst_mode->clock == 594000)
+				rk628_mode_copy(src_mode, dst_mode);
 
 			break;
 		}
@@ -568,6 +571,9 @@ static int rk628_hdmirx_init(struct rk628 *rk628)
 	hdmirx->hpd_output_inverted = of_property_read_bool(dev->of_node,
 		"hpd-output-inverted");
 
+	hdmirx->src_mode_4K_yuv420 = of_property_read_bool(dev->of_node,
+		"src-mode-4k-yuv420");
+
 	/* HDMIRX IOMUX */
 	rk628_i2c_write(rk628, GRF_GPIO1AB_SEL_CON, HIWORD_UPDATE(0x7, 10, 8));
 	//i2s pinctrl
-- 
2.17.1

